### PR TITLE
Modify IViewComponentHelper to remove method selection ambiguity.

### LINKED ISSUE #612: Loosen ViewComponent invoke restrictions
@rynowak and I talked a bit about this in person and I found it very unappealing how in today's world we don't provide a default signature for Invoke/InvokeAsync which makes building ViewComponents inconvenient.

Potential fixes to this could be allowing 
```C#
Component.InvokeAsync("typeName", "methodName", myparams)
```
Or some derivative. Even possibly:
```C#
Component.Invoke<MyParamObject>("typeName", new MyParamObject(...))
```
Which would allow us to have a virtual Invoke/InvokeAsync signature.

Of course these are just ideas thrown around and neither may be the right solution to the problem.  We should discuss making this API simpler and easier to get started with to prevent users from writing code that never gets invoked.

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #612

### CODE REVIEW

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/ViewComponents/ViewComponentMethodSelectorTest.cs [hunk 1]
```diff
@@ -1,282 +0,0 @@
-﻿// Copyright (c) .NET Foundation. All rights reserved.
```
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/ViewComponents/ViewComponentMethodSelectorTest.cs]: Ignore this for now. I'll revive this.

File: test/WebSites/RazorWebSite/Views/ViewEngine/ViewWithRelativePath.cshtml [hunk 2]
```diff
@@ -12,4 +12,4 @@
 }
 ViewWithRelativePath-content
 <partial>@await Html.PartialAsync("../Shared/_PartialThatSetsTitle.cshtml")</partial>
-@await Component.InvokeAsync("ComponentWithRelativePath", person)
\ No newline at end of file
+@await Component.InvokeAsync("ComponentWithRelativePath", new { person })
```
Comment (pranavkm) [inline test/WebSites/RazorWebSite/Views/ViewEngine/ViewWithRelativePath.cshtml]: This.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/DefaultViewComponentInvoker.cs [hunk 3]
```diff
@@ -179,11 +157,13 @@ public IViewComponentResult InvokeSyncCore(MethodInfo method, ViewComponentConte
             {
                 _diagnosticSource.BeforeViewComponent(context, component);
                 _logger.ViewComponentExecuting(context);
-
+                var indexedArguments = ControllerActionExecutor.PrepareArguments(
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/DefaultViewComponentInvoker.cs]: Should we throw here if any values are null?

File: src/Microsoft.AspNet.Mvc.ViewFeatures/IViewComponentHelper.cs [hunk 4]
```diff
@@ -9,20 +9,12 @@ namespace Microsoft.AspNet.Mvc
...
-        void RenderInvoke(Type componentType, params object[] args);
-
-        Task<IHtmlContent> InvokeAsync(string name, params object[] args);
-
-        Task<IHtmlContent> InvokeAsync(Type componentType, params object[] args);
-
-        Task RenderInvokeAsync(string name, params object[] args);
-
-        Task RenderInvokeAsync(Type componentType, params object[] args);
+        Task RenderInvokeAsync(Type componentType, object arguments);
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/IViewComponentHelper.cs]: https://github.com/aspnet/Mvc/issues/612#issuecomment-159103197

No more `RenderInvokeAsync

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponentHelperExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,21 @@
+﻿using System;
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponentHelperExtensions.cs]: license and docs yea.

YEAHIKNOWITSADESIGNPRANDIDONTCARE

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/DefaultViewComponentInvoker.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/DefaultViewComponentInvoker.cs]: We should make sure to support default values as well. 

I don't know what you mean? why would null be invalid?

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/ViewComponentContext.cs [hunk 6]
```diff
@@ -87,7 +81,7 @@ public ViewComponentContext()
         /// <remarks>
         /// The property setter is provided for unit test purposes only.
         /// </remarks>
-        public object[] Arguments { get; set; }
+        public object Arguments { get; set; }
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/ViewComponentContext.cs]: This should be a dictionary

Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/ViewComponentContext.cs]: See ActionFilterContext

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/ViewComponentMethodSelector.cs [hunk 7]
```diff
@@ -63,36 +62,5 @@ public static MethodInfo FindSyncMethod(TypeInfo componentType, object[] args)
 
             return method;
         }
-
-        private static MethodInfo GetMethod(TypeInfo componentType, object[] args, string methodName)
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/ViewComponentMethodSelector.cs]: :clap:

File: test/WebSites/RazorWebSite/Views/ViewEngine/ViewWithRelativePath.cshtml (hunk 2 above)
Comment (rynowak) [inline test/WebSites/RazorWebSite/Views/ViewEngine/ViewWithRelativePath.cshtml]: :heart_eyes_cat:

**Comment by rynowak** [comment]
This will sooo need an announcement

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/DefaultViewComponentInvoker.cs (hunk 3 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/DefaultViewComponentInvoker.cs]: In the event an argument does not have a value?

Component.InvokeAsync("Foo", new { a = "some value" }) -->  InvokeAsync(object a, object b)

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/ViewComponentContext.cs (hunk 6 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/ViewComponentContext.cs]: Ok

File: src/Microsoft.AspNet.Mvc.ViewFeatures/IViewComponentHelper.cs (hunk 4 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/IViewComponentHelper.cs]: Ahh. It sounded like we were getting rid of the synchronous variants.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/DefaultViewComponentInvoker.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/DefaultViewComponentInvoker.cs]: >In the event an argument does not have a value?

Yes. The default value needs to get plugged in at that point

File: src/Microsoft.AspNet.Mvc.ViewFeatures/IViewComponentHelper.cs (hunk 4 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/IViewComponentHelper.cs]: Yeah, both of those. We're simplifying it to the minimum needed.

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/ViewComponents/ViewComponentMethodSelectorTest.cs (hunk 1 above)
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/ViewComponents/ViewComponentMethodSelectorTest.cs]: The code formerly referred to as `ViewComponentMethodSelector` is part of the descriptor provider now.

**Comment by pranavkm** [comment]
:up: :date:

File: test/WebSites/RazorWebSite/Views/ViewEngine/ViewWithRelativePath.cshtml (hunk 2 above)
Comment (pranavkm) [inline test/WebSites/RazorWebSite/Views/ViewEngine/ViewWithRelativePath.cshtml]: Not for this PR, but we could try to be clever in this case and attempt to make it work if the user does

`@await Component.InvokeAsync("ComponentWithRelativePath", person)` (similarly with a params array). The user experience for single parameters is a lot worse than what it used to be before.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/IViewComponentHelper.cs [hunk 8]
```diff
@@ -7,22 +7,27 @@
...
-
-        Task<IHtmlContent> InvokeAsync(string name, params object[] args);
-
-        Task<IHtmlContent> InvokeAsync(Type componentType, params object[] args);
-
-        Task RenderInvokeAsync(string name, params object[] args);
-
-        Task RenderInvokeAsync(Type componentType, params object[] args);
+        /// <summary>
+        /// Invokes a View Component with the specified <paramref name="name"/>.
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/IViewComponentHelper.cs]: Is the casing for this `View Component` or `view component`? I saw some docs earlier which used the former and stuck with it.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Logging/ViewComponentResultLoggerExtensions.cs [hunk 9]
```diff
@@ -18,32 +19,35 @@ static ViewComponentResultLoggerExtensions()
...
-                {
-                    formattedArguments[i] = Convert.ToString(arguments[i]);
-                }
-
+                var formattedArguments = GetFormattedArguments(arguments);
                 _viewComponentResultExecuting(logger, viewComponentType.Name, formattedArguments, null);
             }
         }
+
+        private static string[] GetFormattedArguments(object arguments)
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Logging/ViewComponentResultLoggerExtensions.cs]: @rynowak should we get rid of this (i.e. remove the arguments from logging?)

File: src/Microsoft.AspNet.Mvc.ViewFeatures/IViewComponentHelper.cs [hunk 10]
```diff
@@ -7,22 +7,27 @@
...
+        /// </summary>
+        /// <param name="name">The name of the View Component.</param>
+        /// <param name="arguments">Arguments to be passed to tbe invoked View Component method.</param>
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/IViewComponentHelper.cs]: `the`

File: src/Microsoft.AspNet.Mvc.ViewFeatures/IViewComponentHelper.cs [hunk 11]
```diff
@@ -7,22 +7,27 @@
...
+        /// <returns>A <see cref="Task"/> that on completion returns the rendered <see cref="IHtmlContent" />.
+        /// </returns>
+        Task<IHtmlContent> InvokeAsync(string name, object arguments);
+
+        /// <summary>
+        /// Invokes a View Component of type <paramref name="componentType" />.
+        /// </summary>
+        /// <param name="componentType">The View Component <see cref="Type"/>.</param>
+        /// <param name="arguments">Arguments to be passed to tbe invoked View Component method.</param>
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/IViewComponentHelper.cs]: `the`

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Logging/DefaultViewComponentInvokerLoggerExtensions.cs [hunk 12]
```diff
@@ -40,10 +40,10 @@ public static void ViewComponentExecuting(this ILogger logger, ViewComponentCont
         public static void ViewComponentExecuted(
             this ILogger logger,
             ViewComponentContext context,
-            int startTime,
+            DateTime startTime,
             object result)
         {
-            var elapsed = new TimeSpan(Environment.TickCount - startTime);
+            var elapsed = DateTime.UtcNow - startTime;
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Logging/DefaultViewComponentInvokerLoggerExtensions.cs]: Do we care about accuracy here? Doing date time comparisons to compute elapsed times isn't very accurate.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Logging/ViewComponentResultLoggerExtensions.cs [hunk 13]
```diff
@@ -18,32 +19,35 @@ static ViewComponentResultLoggerExtensions()
...
+        {
+            var argumentDictionary = PropertyHelper.ObjectToDictionary(arguments);
+            var formattedArguments = new string[argumentDictionary.Count];
+            var i = 0;
+            foreach (var item in argumentDictionary)
+            {
+                formattedArguments[i++] = Convert.ToString(item.Value);
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Logging/ViewComponentResultLoggerExtensions.cs]: I know we didn't do this before but we don't care about the item keys at all?

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/ViewComponentHelperExtensions.cs [hunk 14]
```diff
@@ -7,48 +7,76 @@
...
         }
 
-        public static Task<IHtmlContent> InvokeAsync<TComponent>(
-            this IViewComponentHelper helper,
-            params object[] args)
+        /// <summary>
+        /// Invokes a View Component of type <typeparam name="TComponent"/>.
+        /// </summary>
+        /// <param name="helper">The <see cref="IViewComponentHelper"/>.</param>
+        /// <param name="arguments">Arguments to be passed to tbe invoked View Component method.</param>
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/ViewComponentHelperExtensions.cs]: `the`

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Logging/DefaultViewComponentInvokerLoggerExtensions.cs (hunk 12 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Logging/DefaultViewComponentInvokerLoggerExtensions.cs]: It's better than using `Environment.TickCounts` according to http://stackoverflow.com/a/8865560. I think the reason it doesn't use a `Stopwatch` is to avoid the allocation for logging.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/ViewComponentContext.cs [hunk 15]
```diff
@@ -48,11 +48,6 @@ public ViewComponentContext()
                 throw new ArgumentNullException(nameof(viewComponentDescriptor));
             }
 
-            if (arguments == null)
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/ViewComponentContext.cs]: Undo

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Resources.resx [hunk 16]
```diff
@@ -137,10 +137,7 @@
     <value>An invoker could not be created for the view component '{0}'.</value>
   </data>
   <data name="ViewComponent_CannotFindMethod" xml:space="preserve">
-    <value>Could not find an '{0}' method matching the parameters.</value>
-  </data>
-  <data name="ViewComponent_CannotFindMethod_WithFallback" xml:space="preserve">
-    <value>Could not find an '{0}' or '{1}' method matching the parameters.</value>
+    <value>Could not find an '{0}' or '{1}' method for the view component '{2}'.</value>
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Resources.resx]: Should stay consistent with elsewhere when saying `view component` vs. `View Component`

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/DefaultViewComponentDescriptorProvider.cs [hunk 17]
```diff
@@ -14,6 +16,8 @@ namespace Microsoft.AspNet.Mvc.ViewComponents
     /// </summary>
     public class DefaultViewComponentDescriptorProvider : IViewComponentDescriptorProvider
     {
+        private const string AsyncMethodName = "InvokeAsync";
+        private const string SyncMethodName = "Invoke";
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/DefaultViewComponentDescriptorProvider.cs]: So it's always `@Component.InvokeAsync` but you can write your `ViewComponent` with `Invoke` OR `InvokeAsync`? Interesting.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/DefaultViewComponentDescriptorProvider.cs [hunk 18]
```diff
@@ -63,16 +67,69 @@ protected virtual bool IsViewComponentType(TypeInfo typeInfo)
...
             };
 
             return candidate;
         }
+
+        private static MethodInfo FindMethod(Type componentType)
+        {
+            var componentName = componentType.FullName;
+            var methods = componentType.GetMethods(BindingFlags.Public | BindingFlags.Instance)
+                .Where(m => string.Equals(m.Name, AsyncMethodName, StringComparison.Ordinal) ||
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/DefaultViewComponentDescriptorProvider.cs]: `m` => `method` who cares about long line here imo.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Resources.resx [hunk 19]
```diff
@@ -122,13 +122,13 @@
...
   <data name="ViewComponent_AsyncMethod_ShouldReturnTask" xml:space="preserve">
-    <value>The async view component method '{0}' should be declared to return Task&lt;T&gt;.</value>
+    <value>Method '{0}' of view component '{1}' should be declared to return {2}&lt;T&gt;.</value>
   </data>
   <data name="ViewComponent_MustReturnValue" xml:space="preserve">
     <value>A view component must return a non-null value.</value>
   </data>
   <data name="ViewComponent_SyncMethod_ShouldReturnValue" xml:space="preserve">
-    <value>The view component method '{0}' should be declared to return a value.</value>
+    <value>Method '{0}' of view component '{1}' should be declared to return a value.</value>
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Resources.resx]: Could add a comment saying an alternative (if you don't want to return a value) is to inject services.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/DefaultViewComponentInvoker.cs [hunk 20]
```diff
@@ -55,60 +65,33 @@ public class DefaultViewComponentInvoker : IViewComponentInvoker
...
             if (context == null)
             {
                 throw new ArgumentNullException(nameof(context));
             }
 
-            var method = ViewComponentMethodSelector.FindSyncMethod(
-                context.ViewComponentDescriptor.Type.GetTypeInfo(),
-                context.Arguments);
-            if (method == null)
+            var methodInfo = context.ViewComponentDescriptor?.MethodInfo;
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/DefaultViewComponentInvoker.cs]: `ViewComponentDescriptor` can be `null`?

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/ViewComponentContext.cs (hunk 15 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/ViewComponentContext.cs]: Arguments still can't be `null` can it? The `PropertyHelper.ObjectToDictionary` should always at least create a new dictionary for you.

File: test/WebSites/RazorWebSite/Views/ViewEngine/ViewWithRelativePath.cshtml (hunk 2 above)
Comment (NTaylorMullen) [inline test/WebSites/RazorWebSite/Views/ViewEngine/ViewWithRelativePath.cshtml]: One thing, tooling sucks with this doesn't it? Aka second you type `{` does it properly auto-complete the braces? If not we should file a tooling bug. I think I recall it being difficult to type out in the Razor editor.

**Comment by NTaylorMullen** [comment]
:watch:

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/ViewComponentContext.cs (hunk 15 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/ViewComponentContext.cs]: Yeah, it can't be null anymore.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/IViewComponentHelper.cs (hunk 8 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/IViewComponentHelper.cs]: I think I've generally done lower case.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Logging/ViewComponentResultLoggerExtensions.cs (hunk 9 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Logging/ViewComponentResultLoggerExtensions.cs]: No

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Logging/DefaultViewComponentInvokerLoggerExtensions.cs (hunk 12 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Logging/DefaultViewComponentInvokerLoggerExtensions.cs]: We don't care about accuracy beyond the ms level. What we're doing here is intentional

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Logging/ViewComponentResultLoggerExtensions.cs (hunk 13 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Logging/ViewComponentResultLoggerExtensions.cs]: As long as they are in the correct order, we'd want the keys... It looks like they are not in the correct order.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Resources.resx (hunk 19 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Resources.resx]: I don't need this comment. `void` and `Task` are specifically not allowed for VCs

File: src/Microsoft.AspNet.Mvc.ViewFeatures/IViewComponentHelper.cs (hunk 8 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/IViewComponentHelper.cs]: Yeah let's just standardize on lower case

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Microsoft.AspNet.Mvc.ViewFeatures.xproj [hunk 21]
```diff
@@ -3,6 +3,7 @@
   <PropertyGroup>
     <VisualStudioVersion Condition="'$(VisualStudioVersion)' == ''">14.0</VisualStudioVersion>
     <VSToolsPath Condition="'$(VSToolsPath)' == ''">$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)</VSToolsPath>
+    <DisabledCustomTools>.resx</DisabledCustomTools>
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Microsoft.AspNet.Mvc.ViewFeatures.xproj]: :heart_eyes_cat:

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Logging/DefaultViewComponentInvokerLoggerExtensions.cs (hunk 12 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Logging/DefaultViewComponentInvokerLoggerExtensions.cs]: Wait, I just noticed the change. If you're going to make this change, do it more globablly and not part of this PR. We have 3-4 different places we do the same thing in code.

**Comment by rynowak** [comment]
:watch:

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Logging/DefaultViewComponentInvokerLoggerExtensions.cs (hunk 12 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Logging/DefaultViewComponentInvokerLoggerExtensions.cs]: Okie dokie.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Logging/ViewComponentResultLoggerExtensions.cs (hunk 13 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Logging/ViewComponentResultLoggerExtensions.cs]: We don't have the descriptor at this point to figure out what the order is. Are there any other places that log a dictionary?

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Resources.resx (hunk 19 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Resources.resx]: :+1: to @rynowak. You can't have a ViewComponent Invoke method that returns nothing. (which is pretty much what the exception message says)

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Resources.resx (hunk 16 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Resources.resx]: `view component` it is.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/DefaultViewComponentInvoker.cs (hunk 20 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/ViewComponents/DefaultViewComponentInvoker.cs]: It's settable and we only verify if the context is non null at the start of the method.

File: test/WebSites/RazorWebSite/Views/ViewEngine/ViewWithRelativePath.cshtml (hunk 2 above)
Comment (pranavkm) [inline test/WebSites/RazorWebSite/Views/ViewEngine/ViewWithRelativePath.cshtml]: Tooling works fine.

**Comment by pranavkm** [comment]
Updated

**Comment by NTaylorMullen** [comment]
:shipit:

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Logging/ViewComponentResultLoggerExtensions.cs (hunk 13 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Logging/ViewComponentResultLoggerExtensions.cs]: Look at what we do for actions. It might be ok to log keys + values.

**Comment by pranavkm** [comment]
@rynowak moved the logging to the invoker where we have ordered arguments.
